spec.md

```diff
@@ -570,13 +559,297 @@ schema being used.
         "bullets": [
           "40 concurrent connections"
         ]
+      },
+      "schemas": {
+        "service_instance": {
+          "create": {
+            "parameters": {
+              "$ref": "osb_v2:///fakeservice/v1#BillingAccountParameters"
+            }
+          },
+          "update": {
+            "parameters": {
+              "$schema": "http://json-schema.org/draft-04/schema#",
+              "type": "object",
+              "propeties": {
+                "some-other-account": {
+                  "description": "Other fake account needed for update.",
+                  "type": "string"
+                },
+                "billing": {
+                  "$ref": "osb_v2:///fakeservice/v1#BillingAccountParameters"
+                }
+              }
+            }
+          }
+        },
+        "service_binding": {
+          "create": {
+            "parameters": {
+              "$ref": "osb_v2:///fakeservice/v1/bindings#AccountParameters"
+            }
+          }
+        }
       }
     }]
   }]
 }
 ```
 
 
+## JSON Schemas
+
+The JSON Schemas endpoint is used to define reusable definitions of JSON Schema
+Objects referenced throughout the [Catalog](#catalog-management). If a Platform
+intends to support JSON Schemas, it MUST examine the response from the catalog
+to determine if the Service Broker is using `$ref` JSON Schemas. The Service
+Broker can signal to the Platform that JSON Schema components are supported by
+using the `osb_v2:` scheme. Components are a way to allow Platforms to only
+fetch sub-schemas.
+
+The root JSON Schema document can be found by getting the JSON Schemas
+endpoint without a `:component_path`.
+
+If the `osb_v2:` scheme is used by the Service Broker, then the Platform MAY
+fetch the component JSON Schema directly using the following convention:
+
+ - Given: `osb_v2:///{:component_path}`, fetch
+   `/v2/catalog/schemas/:component_path` and then the Platform is responsible
+   to continue fetching any unknown `:component_path` URIs. Care ought to be
+   taken by the platform to not fetch duplicate definitions.
+
+A Platform MAY ignore the `osb_v2:` scheme and just fetch `/v2/catalog/schemas`
+which will contain all sub-Schema documents. This reduces the complexity of the
+Platform but increases the potential payload size of the JSON Schemas.
+
+A Platform SHOULD fetch the root Schema document for any other `$ref` URIs
+found that do not match the `osb_v2:` scheme. The Service Broker MUST provide
+the JSON Schema Object for the referenced URI in a subschema returned.
+
+Note: If the Service Broker provides a mix of `osb_v2:` in addition to other
+schemes that result in fetching both `/v2/catalog/schemas/:component_path` and
+`/v2/catalog/schemas`, the Platform will have colliding JSON Schema `$ids`. In
+this case the Platform SHOULD ignore the component paths and use the root
+schema. Service Brokers SHOULD avoid mixing `osb_v2:` and other schemes.
```

Comment:
re: https://github.com/openservicebrokerapi/servicebroker/pull/402#pullrequestreview-92616345

I agree its hard but we need to say something about this. If a Broker is changing its service definitions often it has no requirement to keep the old schemas around. Which means that if the interval that a Platform has for re-listing the services is longer than the interval at which a Broker updates (and removes) its schema then the Platform will end up missing data.

Perhaps for now the best we can do is say that if the Broker returns a 404/410? from a schema GET then it SHOULD re-list since it probably means the Broker's service list and schema have changed.  Although, an existing service usage is now in trouble.  Maybe we need to say that AND that a Broker can't remove a schema as long as there is an existing instance?

Also, we should be clear on what a Broker returns when a schema is removed - 404 or 410? Should probably be a 410 to be consistent with the rest of the spec.

And we should probably say that an existing schema reference can only be updated as long as its done so in a backward compatible way.

Reply:
>AND that a Broker can't remove a schema as long as there is an existing instance?

I can see a service class add a parameter that would be in support of the next minor version of that class, it is optional and therefor not breaking of the existing instance. The suggestion would block this behavior and that is not ideal.

> 410 to be consistent with the rest of the spec.

This means the broker needs to keep track of all schemas ever offered. I don't want to sign the broker authors up for this. 

I think we need some real implementations to test if the platform can get easily confused. I suspect the issues you are raising here are non-issues for versatile platforms. The issues at the root are pull-based APIs.